webui/react/src/components/kit/Modal.tsx

```diff
@@ -0,0 +1,192 @@
+  createContext,
+  Dispatch,
+  ReactNode,
+  SetStateAction,
+  useCallback,
+  useContext,
+  useState,
+} from 'react';
+
+import Button from 'components/kit/Button';
+import Link from 'components/Link';
+import Icon from 'shared/components/Icon';
+import Spinner from 'shared/components/Spinner';
+import { ErrorLevel, ErrorType } from 'shared/utils/error';
+import handleError from 'utils/error';
+
+import css from './Modal.module.scss';
+
+interface LinkParams {
+  text: string;
+  url: string;
+}
+
+export type ModalSize = 'small' | 'medium' | 'large';
+const modalWidths: { [key in ModalSize]: number } = {
+  large: 1025,
+  medium: 692,
+  small: 358,
+};
+
+export type Opener = Dispatch<SetStateAction<boolean>>;
+
+export type ModalContext = {
+  isOpen: boolean;
+  setIsOpen: Opener;
+};
+
+export interface ModalSubmitParams {
+  disabled?: boolean;
+  text: string;
+  handler: () => Promise<void>;
+  onComplete?: () => Promise<void>;
+}
+
+interface ModalProps {
+  cancel?: boolean;
+  cancelText?: string;
+  danger?: boolean;
+  footerLink?: LinkParams;
+  headerLink?: LinkParams;
+  icon?: string;
+  key?: string;
+  size?: ModalSize;
+  submit?: ModalSubmitParams;
+  title: string;
+  children: ReactNode;
+}
+
+const ModalContext = createContext<ModalContext | null>(null);
+
+export const Modal: React.FC<ModalProps> = ({
+  cancel,
+  cancelText,
+  danger,
+  footerLink,
+  headerLink,
+  icon,
+  key,
+  size = 'large',
+  submit,
+  title,
+  children: modalBody,
+}: ModalProps) => {
+  const modalContext = useContext(ModalContext);
+
+  if (modalContext === null) {
+    throw new Error('Modal used outside of ModalContext');
+  }
+  const { isOpen, setIsOpen } = modalContext;
+
+  const [isSubmitting, setIsSubmitting] = useState(false);
+
+  const close = useCallback(() => setIsOpen(false), [setIsOpen]);
+
+  const handleSubmit = useCallback(async () => {
+    setIsSubmitting(true);
+    try {
+      await submit?.handler();
+      setIsSubmitting(false);
+      setIsOpen(false);
+      await submit?.onComplete?.();
+    } catch (err) {
+      handleError(err, {
+        level: ErrorLevel.Error,
+        publicMessage: err instanceof Error ? err.message : '',
+        publicSubject: 'Could not submit form',
+        silent: false,
+        type: ErrorType.Server,
+      });
+      setIsSubmitting(false);
+    }
+  }, [submit, setIsOpen]);
+
+  return (
+    <AntdModal
+      cancelText={cancelText}
+      footer={
+        (footerLink || cancel || submit) ? (
+          <div className={css.footer}>
+            <div>
+              {footerLink && (
+                <Link path={footerLink.url} popout>
+                  {footerLink.text}
+                </Link>
+              )}
+            </div>
+            <div>
+              {(cancel || cancelText) && (
+                <Button key="back" onClick={close}>
+                  {cancelText || 'Cancel'}
+                </Button>
+              )}
+              {submit && (
```

Comment:
Oh, this is a good callout. Can you add a required input to the modal in the UI Kit?

Reply:
One design note: there should be a tooltip on the disabled submit button explaining why it is disabled. Ideally this would aggregate the validation errors in the form but I know AntD's form API isn't super robust so if you can't do that can it can say "Address validation errors before proceeding".